File: personalizatio-sdk/src/main/java/com/personalizatio/stories/models/elements/ProductsElement.java
package com.personalizatio.stories.models.elements;

import androidx.annotation.NonNull;

import com.personalizatio.Product;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ProductsElement implements Element {

    private String labelHide;
    private String labelShow;
    private final List<Product> products = new ArrayList<>();

    public ProductsElement(@NonNull JSONObject json) {
        var labelsJson = json.optJSONObject("labels");
        if (labelsJson != null) {
            labelHide = json.optString("hide_carousel", "");
            labelShow = json.optString("show_carousel", "");
        }
        var productsJsonArray = json.optJSONArray("products");
        if (productsJsonArray != null)
        {
            for (int i = 0; i < productsJsonArray.length(); i++) {
                products.add(new Product(productsJsonArray.optJSONObject(i)));
            }
        }
    }

    public String getLabelHide() {
        return labelHide;
    }

    public String getLabelShow() {
        return labelShow;
    }

    public List<Product> getProducts() {
        return products;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ProductsElement that)) return false;
        return labelHide.equals(that.labelHide)
                && labelShow.equals(that.labelShow)
                && products.equals(that.products);
    }

    @Override
    public int hashCode() {
        return Objects.hash(labelHide, labelShow, products);
    }

    @NonNull
    @Override
    public String toString() {
        return "ProductsElement{" +
                "labelHide='" + labelHide + '\'' +
                ", labelShow='" + labelShow + '\'' +
                ", products=" + products +
                '}';
    }
}

File: personalization-sdk/src/main/kotlin/com/personalization/stories/models/elements/ProductsElement.kt
package com.personalization.stories.models.elements

import com.personalization.Product
import java.util.Objects
import org.json.JSONObject

class ProductsElement(json: JSONObject) : Element {
    var labelHide: String = ""
    var labelShow: String = ""
    private val products: MutableList<Product> = ArrayList()

    init {
        // The backend nests both captions inside a "labels" object, so they have to be read from
        // it rather than from the element itself — reading the parent silently yields empty text
        // and the carousel button renders blank.
        val labelsJson = json.optJSONObject("labels")
        if (labelsJson != null) {
            labelHide = labelsJson.optString("hide_carousel", "")
            labelShow = labelsJson.optString("show_carousel", "")
        }
        val productsJsonArray = json.optJSONArray("products")
        if (productsJsonArray != null) {
            for (i in 0 until productsJsonArray.length()) {
                products.add(Product(productsJsonArray.optJSONObject(i)))
            }
        }
    }

    fun getProducts(): List<Product> {
        return products
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ProductsElement) return false
        return labelHide == other.labelHide
            && labelShow == other.labelShow
            && products == other.products
    }

    override fun hashCode(): Int {
        return Objects.hash(labelHide, labelShow, products)
    }

    override fun toString(): String {
        return "ProductsElement{labelHide='$labelHide', labelShow='$labelShow', products=$products}"
    }
}
